vpr/src/route/route_common.cpp

```diff
@@ -307,14 +307,33 @@ bool try_route(int width_fac,
         IntraLbPbPinLookup intra_lb_pb_pin_lookup(device_ctx.logical_block_types);
         ClusteredPinAtomPinsLookup netlist_pin_lookup(cluster_ctx.clb_nlist, intra_lb_pb_pin_lookup);
 
-        success = try_timing_driven_route(router_opts,
-                                          analysis_opts,
-                                          segment_inf,
-                                          net_delay,
-                                          netlist_pin_lookup,
-                                          timing_info,
-                                          delay_calc,
-                                          first_iteration_priority);
+        switch (router_opts.router_heap) {
+            case e_heap_type::BINARY_HEAP:
+                success = try_timing_driven_route<BinaryHeap>(
+                    router_opts,
+                    analysis_opts,
+                    segment_inf,
+                    net_delay,
+                    netlist_pin_lookup,
+                    timing_info,
+                    delay_calc,
+                    first_iteration_priority);
+                break;
+            case e_heap_type::BUCKET_HEAP_APPROXIMATION:
+                success = try_timing_driven_route<Bucket>(
+                    router_opts,
+                    analysis_opts,
+                    segment_inf,
+                    net_delay,
+                    netlist_pin_lookup,
+                    timing_info,
+                    delay_calc,
+                    first_iteration_priority);
+                break;
+            default:
+                VPR_ERROR(VPR_ERROR_ROUTE, "Unknown heap_type %d",
+                          router_opts.router_heap);
+        }
```

Comment:
This is a bit awkward, and requires templating all the timing_driven_route_* functions by the heap type.

What are you're thoughts about making a `ConnectionRouterInterface`, and having `ConnectionRouter<BinaryHeap>` and `ConnectionRouter<Bucket>` implement it?

That would ensure the core loop (path search) is fully templated, while avoiding templating the rest of the router. 

Since the number of invocations of the Connection router is  small compared to the number of heap operations any overhead of the virtual call should be negligible (e.g. on the large Titan bitcoin_miner design, there are 178x fewer invocations of the connection router than heap operations (3M vs 680M)).

Reply:
I'll post the virtual version soon and once the nightly CI completes we can compare https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1185/commits/87577fd21f8865477607afd85428702d6334667c (fully templated) with the new interface-at-ConnectionRouter.

If the performance hit is small, I think having the `ConnectionRouterInterface` interface is superior to fully templating the router!

Reply:
I've kicked off two kokoro runs:
 - Nightly with fully templated router: https://source.cloud.google.com/results/invocations/4ceccd64-1170-42f7-be59-b816c53300fe
 - Nightly with `ConnectionRouterInterface`: https://source.cloud.google.com/results/invocations/0ed2f9b6-f74a-4824-af6d-ef9c25e10c62/targets

In about 12-16 hrs we'll be able to compare the relative impacts.

Reply:
I like the virtualized interface, it keeps the code much cleaner.

We'll see how the results come back, but hopefully the impacts will be small!